### ISSUE TITLE
cmd/go: get: panic: internal error: can't find reason for requirement on golang.org/x/sync@...

### ISSUE THREAD

**Issue by icio** [golang/go#56494]
### What version of Go are you using (`go version`)?

<pre>
$ gotip version
go version devel go1.20-e09bbae Sat Oct 29 04:48:07 2022 +0000 linux/amd64
</pre>

Same result with 1.18.6, 1.19.

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/icio/.cache/go-build"
GOENV="/home/icio/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/icio/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/icio/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/icio/sdk/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/icio/sdk/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="devel go1.20-e09bbae Sat Oct 29 04:48:07 2022 +0000"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/icio/go/src/github.com/icio/go-get-panic/go.mod"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2460736486=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Run `go get -u golang.org/x/tools` with go.mod:

```go.mod
module github.com/icio/go-get-panic

go 1.18

require (
	golang.org/x/net v0.1.0
	golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f
	golang.org/x/text v0.4.0
)
```

### What did you expect to see?

go.mod updated with latest golang.org/x/tools version.

### What did you see instead?

```
panic: internal error: can't find reason for requirement on golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0xc0000281e0, 0x11}, {0xc000032690, 0x22}})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1760 +0xd4
cmd/go/internal/modget.(*resolver).updateBuildList(0xc00023c000, {0xb2d5d0, 0xc00002c0f8}, {0x0, 0x0, 0x0})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1765 +0x54c
cmd/go/internal/modget.(*resolver).resolveQueries(0xc00023c000, {0xb2d5d0, 0xc00002c0f8}, {0xc000014058, 0x1, 0xc0000a3d78?})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1243 +0x1a5
cmd/go/internal/modget.runGet({0xb2d5d0, 0xc00002c0f8}, 0xc0000285e8?, {0xc0000240f0, 0x1, 0x1})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:314 +0x408
main.invoke(0xe371a0, {0xc0000240d0, 0x3, 0x3})
        /home/icio/sdk/gotip/src/cmd/go/main.go:225 +0x3d9
main.main()
        /home/icio/sdk/gotip/src/cmd/go/main.go:179 +0x7ce
```

### To reproduce

```
$ git clone https://github.com/icio/go-get-panic
Cloning into 'go-get-panic'...
remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 7 (delta 1), reused 7 (delta 1), pack-reused 0
Receiving objects: 100% (7/7), done.
Resolving deltas: 100% (1/1), done.

$ cd go-get-panic

$ git checkout 2cbc2db2a48105fa2b71ec69d11b6496ba52c263

$ git rev-parse HEAD
2cbc2db2a48105fa2b71ec69d11b6496ba52c263

$ gotip version
go version devel go1.20-e09bbae Sat Oct 29 04:48:07 2022 +0000 linux/amd64

$ gotip get -u golang.org/x/tools
panic: internal error: can't find reason for requirement on golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0xc0000281e0, 0x11}, {0xc000032690, 0x22}})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1760 +0xd4
cmd/go/internal/modget.(*resolver).updateBuildList(0xc00023c000, {0xb2d5d0, 0xc00002c0f8}, {0x0, 0x0, 0x0})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1765 +0x54c
cmd/go/internal/modget.(*resolver).resolveQueries(0xc00023c000, {0xb2d5d0, 0xc00002c0f8}, {0xc000014058, 0x1, 0xc0000a3d78?})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:1243 +0x1a5
cmd/go/internal/modget.runGet({0xb2d5d0, 0xc00002c0f8}, 0xc0000285e8?, {0xc0000240f0, 0x1, 0x1})
        /home/icio/sdk/gotip/src/cmd/go/internal/modget/get.go:314 +0x408
main.invoke(0xe371a0, {0xc0000240d0, 0x3, 0x3})
        /home/icio/sdk/gotip/src/cmd/go/main.go:225 +0x3d9
main.main()
        /home/icio/sdk/gotip/src/cmd/go/main.go:179 +0x7ce
```

This reproduces even if you remove the go.sum:

```
$ rm go.sum && gotip get -u golang.org/x/tools 2>&1 | head -1
panic: internal error: can't find reason for requirement on golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f
```

Uncommenting any of the requirements in go.mod and re-running `go get -u golang.org/x/tools` will allow `go get` to succeed. E.g.:

```
$ sed -i '/x\/net/d' go.mod  

$ git diff
diff --git a/go.mod b/go.mod
index cc0392c..05b38b1 100644
--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,6 @@ module github.com/icio/go-get-panic
 go 1.18
 
 require (
-       golang.org/x/net v0.1.0
        golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f
        golang.org/x/text v0.4.0
 )

$ go get -u golang.org/x/tools
go: downgraded golang.org/x/text v0.4.0 => v0.3.7
go: upgraded golang.org/x/tools v0.1.12 => v0.2.0
```

In the repository where I first encountered, this I was able to fix `go get -u golang.org/x/tools` by removing the `golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f` line. Next, re-running `go mod tidy` reintroduces `golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f` to go.mod. I'm not sure if this is related, but despite having a ./go.mod at the root of the project, we also have a ./cmd/ci-test-utility/go.mod which contains `golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f`.

**Comment by icio** [comment]
There's a few other people tripping over this in https://github.com/golang/go/issues/47979

**Comment by bcmills** [comment]
I see exactly the behavior that you describe. Thank you for distilling it down to a simple reproducer!

**Comment by bcmills** [comment]
The conflict in this example is:
```
go: golang.org/x/tools@upgrade (v0.2.0) requires golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4, not (INTERNAL ERROR: no reason found for golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f)
```

For some reason `modload.editRequirements` things that `golang.org/x/tools@v0.2.0` requires `golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4`, but the limits haven't been updated to allow that version, and with good reason: with graph pruning in effect, `golang.org/x/tools@v0.2.0` doesn't depend on `golang.org/x/sync` *at all*:

```
~/x/tools$ git checkout v0.2.0
HEAD is now at f112c4332 go.mod: update golang.org/x dependencies

~/x/tools$ go list -m golang.org/x/sync
go: module golang.org/x/sync: not a known dependency
```

So I think this is another symptom of #55955.

**Comment by bcmills** [comment]
This arises from a dependency cycle that should be pruned out but isn't:
```
golang.org/x/tools@v0.2.0 golang.org/x/net@v0.1.0
golang.org/x/net@v0.1.0 golang.org/x/text@v0.4.0
golang.org/x/text@v0.4.0 golang.org/x/tools@v0.1.12
golang.org/x/tools@v0.1.12 golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4
```

The cycle isn't pruned because each of the module paths involved in the cycle appears explicitly in the reproducer's `go.mod` file, which causes those dependencies to be followed (because as far as the `versionLimiter` knows at that point they could become selected).

**Comment by godcong** [comment]
I tried the temporary solution to force the version number of sync to solve the problem.
```
replace golang.org/x/sync v0.0.0-20210220032951-036812b2e83c => golang.org/x/sync v0.1.0
```

**Comment by bcmills** [comment]
@godcong, the best temporary workaround is probably to just manually update the `require` line for `golang.org/x/sync` to the version you want and then run `go mod tidy` to clean it up.

(This bug does not affect `go mod tidy`, only `go get`.)

**Comment by godcong** [comment]
> @godcong, the best temporary workaround is probably to just manually update the line for to the version you want and then run to clean it up.`require``golang.org/x/sync``go mod tidy`
> 
> (This bug does not affect , only .)`go mod tidy``go get`

I have tried this program
I used `go get sync@v0.1.0` to update sync to v0.1.0
but it still gives me an error when implementing `go mod tidy`.
Maybe it is because my main project does not use sync.

**Comment by icio** [comment]
@godcong Removing all mention of golang.org/x/sync from go.mod then running my original `go get` command fixed it all for me.

**Comment by godcong** [comment]
> @godcong Removing all mention of golang.org/x/sync from go.mod then running my original command fixed it all for me.`go get`
I can't delete it, because it's an open source project on github.
I'll just have to wait for the owner of this project to fix it... :(

**Comment by KinaneD** [comment]
Same here.
Version:
```
go version
go version go1.18.8 linux/arm64
```
Installing Cloud Storage:
```
go get cloud.google.com/go/storage
panic: internal error: can't find reason for requirement on github.com/jstemmer/go-junit-report@v0.0.0-20190106144839-af01ea7f8024

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0x4000228450?, 0x40000b1560?}, {0x40001005a0?, 0x4000300950?}})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1760 +0xf8
cmd/go/internal/modget.(*resolver).updateBuildList(0x40001c6000, {0x679140, 0x4000026060}, {0x0, 0x0, 0x0})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1765 +0x454
cmd/go/internal/modget.(*resolver).resolveQueries(0x40001c6000, {0x679140, 0x4000026060}, {0x4000010058, 0x1, 0x563e40?})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1243 +0x17c
cmd/go/internal/modget.runGet({0x679140, 0x4000026060}, 0x4000024480?, {0x40000201d0, 0x1, 0x1})
	/usr/local/go/src/cmd/go/internal/modget/get.go:314 +0x394
main.invoke(0x951a00, {0x40000201c0, 0x2, 0x2})
	/usr/local/go/src/cmd/go/main.go:218 +0x2b0
main.main()
	/usr/local/go/src/cmd/go/main.go:175 +0x788
```

Upgrading go-junit-report did not resolve:
```
go install github.com/jstemmer/go-junit-report/v2@latest
go: downloading github.com/jstemmer/go-junit-report v1.0.0
go: downloading github.com/jstemmer/go-junit-report/v2 v2.0.0
```

**Comment by clarkmcc** [comment]
Same as @KinaneD 

When running
```
go get cloud.google.com/go/storage
```

On 
```
$ go version
go version go1.19.3 darwin/arm64
```

I get the following stack trace
```
$ go get cloud.google.com/go/storage
go: downloading cloud.google.com/go/storage v1.28.0
go: downloading cloud.google.com/go v0.107.0
go: downloading cloud.google.com/go/iam v0.5.0
go: downloading cloud.google.com/go/compute/metadata v0.2.1
go: downloading github.com/googleapis/gax-go/v2 v2.6.0
go: downloading golang.org/x/oauth2 v0.0.0-20221014153046-6fdb5e3db783
go: downloading google.golang.org/api v0.102.0
go: downloading google.golang.org/genproto v0.0.0-20221024183307-1bc688fe9f3e
go: downloading cloud.google.com/go/compute v1.12.1
go: downloading google.golang.org/grpc v1.50.1
go: downloading golang.org/x/net v0.0.0-20221014081412-f15817d10f9b
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.2.0
panic: internal error: can't find reason for requirement on github.com/stretchr/testify@v1.6.1

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0x140005ab580?, 0x140003cce10?}, {0x140001a29a0?, 0x1400029c6e0?}})
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/internal/modget/get.go:1760 +0xf8
cmd/go/internal/modget.(*resolver).updateBuildList(0x14000438200, {0x104b0a510, 0x140001a2000}, {0x0, 0x0, 0x0})
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/internal/modget/get.go:1765 +0x450
cmd/go/internal/modget.(*resolver).applyUpgrades(0x14000438200, {0x104b0a510, 0x140001a2000}, {0x0?, 0x1?, 0x104aef8a0?})
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/internal/modget/get.go:1312 +0xc0
cmd/go/internal/modget.runGet({0x104b0a510, 0x140001a2000}, 0x140001bc348?, {0x1400019e1a0, 0x1, 0x1})
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/internal/modget/get.go:351 +0x3c8
main.invoke(0x104dfe280, {0x1400019e190, 0x2, 0x2})
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/main.go:225 +0x2f8
main.main()
        /opt/homebrew/Cellar/go/1.19.3/libexec/src/cmd/go/main.go:179 +0x7ac

```

**Comment by godcong** [comment]
`go 1.19.4` released, has anyone tried it?
Does this problem still exist?

**Comment by Jorropo** [comment]
@godcong yes https://github.com/ipfs/kubo/pull/9423#issuecomment-1344311073 (I tried with go1.19.4 and e76c87b1915cfbd0d4f22c6401987948af18e8c5)

**Comment by KinaneD** [comment]
Hi @clarkmcc I ended up upgrading the dependency that required the conflicting package.
I recommend you start from there.
Good luck.

**Comment by Jorropo** [comment]
I have bisected the issue to 9a81702b974f9d6e5569f069baaad58a0829c63a.

I have multiple failing packages which are deeply connected in my dependency graph making the workaround of upgrading everything / removing them very unpractical.

**Comment by anacrolix** [comment]
Another example I believe: https://github.com/anacrolix/torrent/issues/792.

**Comment by josue** [comment]
@godcong I do not see that it happens on go1.19.4 linux/amd64:

```bash
docker run -it --rm -w /app golang:1.19.4 bash -c "echo; go version; echo; go mod init test; echo; go get cloud.google.com/go/storage; echo; go mod verify"

go version go1.19.4 linux/amd64

go: creating new go.mod: module test

go: downloading cloud.google.com/go/storage v1.28.1
go: downloading cloud.google.com/go v0.107.0
go: downloading cloud.google.com/go v0.105.0
go: downloading cloud.google.com/go/compute/metadata v0.2.1
go: downloading cloud.google.com/go/iam v0.7.0
go: downloading cloud.google.com/go/compute v1.12.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/googleapis/gax-go/v2 v2.7.0
go: downloading golang.org/x/oauth2 v0.0.0-20221014153046-6fdb5e3db783
go: downloading google.golang.org/api v0.103.0
go: downloading google.golang.org/genproto v0.0.0-20221201164419-0e50fba7f41c
go: downloading google.golang.org/grpc v1.50.1
go: downloading google.golang.org/protobuf v1.28.1
go: downloading go.opencensus.io v0.24.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading google.golang.org/appengine v1.6.7
go: downloading golang.org/x/net v0.0.0-20221014081412-f15817d10f9b
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.2.0
go: downloading github.com/golang/protobuf v1.5.2
go: downloading golang.org/x/sys v0.0.0-20220728004956-3c1f35247d10
go: downloading github.com/google/go-cmp v0.5.9
go: downloading golang.org/x/text v0.4.0
go: added cloud.google.com/go v0.105.0
go: added cloud.google.com/go/compute v1.12.1
go: added cloud.google.com/go/compute/metadata v0.2.1
go: added cloud.google.com/go/iam v0.7.0
go: added cloud.google.com/go/storage v1.28.1
go: added github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: added github.com/golang/protobuf v1.5.2
go: added github.com/google/go-cmp v0.5.9
go: added github.com/google/uuid v1.3.0
go: added github.com/googleapis/enterprise-certificate-proxy v0.2.0
go: added github.com/googleapis/gax-go/v2 v2.7.0
go: added go.opencensus.io v0.24.0
go: added golang.org/x/net v0.0.0-20221014081412-f15817d10f9b
go: added golang.org/x/oauth2 v0.0.0-20221014153046-6fdb5e3db783
go: added golang.org/x/sys v0.0.0-20220728004956-3c1f35247d10
go: added golang.org/x/text v0.4.0
go: added golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: added google.golang.org/api v0.103.0
go: added google.golang.org/appengine v1.6.7
go: added google.golang.org/genproto v0.0.0-20221201164419-0e50fba7f41c
go: added google.golang.org/grpc v1.50.1
go: added google.golang.org/protobuf v1.28.1

all modules verified
```

**Comment by stapelberg** [comment]
I also ran into this issue, and can confirm that it still happens with `go1.19.4` and `go1.20rc2`, using the following steps to reproduce:

```
% mkdir /tmp/standalone
% cd /tmp/standalone
% go mod init standalone
% go get github.com/evcc-io/evcc   
```

The error message is:

```
go: downloading github.com/evcc-io/evcc v0.0.0-20230106164214-dbf13052101f
panic: internal error: can't find reason for requirement on github.com/rogpeppe/go-internal@v1.6.1

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0xc00028fb40, 0x1f}, {0xc000122b70, 0x6}})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1760 +0xd4
cmd/go/internal/modget.(*resolver).updateBuildList(0xc0001d0000, {0xb390f8, 0xc00002a138}, {0x0, 0x0, 0x0})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1765 +0x54c
cmd/go/internal/modget.(*resolver).applyUpgrades(0xc0001d0000, {0xb390f8, 0xc00002a138}, {0x0?, 0x0, 0xc0000c1d70?})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1312 +0x465
cmd/go/internal/modget.runGet({0xb390f8, 0xc00002a138}, 0xc00002c660?, {0xc0000241a0, 0x1, 0x1})
	/usr/local/go/src/cmd/go/internal/modget/get.go:351 +0x458
main.invoke(0xe4e240, {0xc000024190, 0x2, 0x2})
	/usr/local/go/src/cmd/go/main.go:225 +0x3d9
main.main()
	/usr/local/go/src/cmd/go/main.go:179 +0x7ce
```

**Comment by r3g-5z** [comment]
Is there a known workaround? I'm unable to build Caddy on Fedora 37, Fedora Rawhide, and Arch Linux because of this issue. Fedora uses Go 1.19.4 and Arch Linux uses 1.19.5.

**Comment by mholt** [comment]
@r3g-5z There are several known workarounds described above :point_up:

**Comment by elchead** [comment]
I also have this `go get` issue:
```
go get github.com/onsi/ginkgo/v2
panic: internal error: can't find reason for requirement on golang.org/x/crypto@v0.0.0-20220516162934-403b01795ae8
```
but manually updating the require of `golang.org/x/crypto` did the trick.

on Go version 1.19.5

**Comment by francislavoie** [comment]
We're seeing this happen for a lot of users of Caddy trying to build with plugins added. In our case, it fails on `pprof`.

We use our build tool `xcaddy` which is essentially a glorified shell script (written in Go) that just strings a bunch of `go` commands together to make a build.

See details below, basically makes a `main.go`, then `go mod init caddy`, then `go get` for each plugin in order, also including the main Caddy module to ensure plugins don't inadvertently use a more recent version of Caddy (happens sometimes if the plugin's main branch changes its `go.mod` to a beta release of Caddy), then `go mod tidy` and `go build`.

It doesn't get to the `go build` in this case, because it panics during `go get`.

<details>
<summary>Details</summary>

```
$ go version                                                                                                   
go version go1.20 linux/amd64

$ xcaddy build --with github.com/dulli/caddy-wol                                                               
2023/02/09 19:30:56 [INFO] Temporary folder: /tmp/buildenv_2023-02-09-1930.2335173491
2023/02/09 19:30:56 [INFO] Writing main module: /tmp/buildenv_2023-02-09-1930.2335173491/main.go
package main

import (
	caddycmd "github.com/caddyserver/caddy/v2/cmd"

	// plug in Caddy modules here
	_ "github.com/caddyserver/caddy/v2/modules/standard"
	_ "github.com/dulli/caddy-wol"
)

func main() {
	caddycmd.Main()
}
2023/02/09 19:30:56 [INFO] Initializing Go module
2023/02/09 19:30:56 [INFO] exec (timeout=10s): /usr/local/go/bin/go mod init caddy 
go: creating new go.mod: module caddy
go: to add module requirements and sums:
	go mod tidy
2023/02/09 19:30:56 [INFO] Pinning versions
2023/02/09 19:30:56 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v github.com/caddyserver/caddy/v2 
go: downloading github.com/caddyserver/caddy v1.0.5
go: downloading github.com/caddyserver/caddy/v2 v2.6.3
go: added github.com/beorn7/perks v1.0.1
go: added github.com/caddyserver/caddy/v2 v2.6.3
go: added github.com/caddyserver/certmagic v0.17.2
go: added github.com/cespare/xxhash/v2 v2.1.2
go: added github.com/go-task/slim-sprig v0.0.0-20210107165309-348f09dbbbc0
go: added github.com/golang/mock v1.6.0
go: added github.com/golang/protobuf v1.5.2
go: added github.com/google/pprof v0.0.0-20210407192527-94a9f03dee38
go: added github.com/google/uuid v1.3.0
go: added github.com/klauspost/cpuid/v2 v2.2.3
go: added github.com/libdns/libdns v0.2.1
go: added github.com/matttproud/golang_protobuf_extensions v1.0.1
go: added github.com/mholt/acmez v1.0.4
go: added github.com/miekg/dns v1.1.50
go: added github.com/onsi/ginkgo/v2 v2.2.0
go: added github.com/prometheus/client_golang v1.14.0
go: added github.com/prometheus/client_model v0.3.0
go: added github.com/prometheus/common v0.37.0
go: added github.com/prometheus/procfs v0.8.0
go: added github.com/quic-go/qpack v0.4.0
go: added github.com/quic-go/qtls-go1-18 v0.2.0
go: added github.com/quic-go/qtls-go1-19 v0.2.0
go: added github.com/quic-go/qtls-go1-20 v0.1.0
go: added github.com/quic-go/quic-go v0.32.0
go: added go.uber.org/atomic v1.9.0
go: added go.uber.org/multierr v1.6.0
go: added go.uber.org/zap v1.24.0
go: added golang.org/x/crypto v0.5.0
go: added golang.org/x/exp v0.0.0-20221205204356-47842c84f3db
go: added golang.org/x/mod v0.6.0
go: added golang.org/x/net v0.5.0
go: added golang.org/x/sys v0.5.0
go: added golang.org/x/term v0.5.0
go: added golang.org/x/text v0.6.0
go: added golang.org/x/tools v0.2.0
go: added google.golang.org/protobuf v1.28.1
2023/02/09 19:30:58 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v github.com/dulli/caddy-wol github.com/caddyserver/caddy/v2 
go: downloading github.com/dulli/caddy-wol v0.0.0-20220902190116-a8cbd416a163
panic: internal error: can't find reason for requirement on github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0xc000c8d3e0, 0x17}, {0xc00002e7e0, 0x22}})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1760 +0xd4
cmd/go/internal/modget.(*resolver).updateBuildList(0xc000184000, {0xb39158, 0xc0000280c8}, {0x0, 0x0, 0x0})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1765 +0x54c
cmd/go/internal/modget.(*resolver).applyUpgrades(0xc000184000, {0xb39158, 0xc0000280c8}, {0x0?, 0x0, 0xc00010fd70?})
	/usr/local/go/src/cmd/go/internal/modget/get.go:1312 +0x465
cmd/go/internal/modget.runGet({0xb39158, 0xc0000280c8}, 0xc00002c678?, {0xc000024220, 0x2, 0x2})
	/usr/local/go/src/cmd/go/internal/modget/get.go:351 +0x458
main.invoke(0xe4e240, {0xc0000241f0, 0x5, 0x5})
	/usr/local/go/src/cmd/go/main.go:225 +0x3d9
main.main()
	/usr/local/go/src/cmd/go/main.go:179 +0x7ce
2023/02/09 19:31:04 [FATAL] exit status 2
```

</details>

**Comment by stevengpn** [comment]
> Hi @clarkmcc I ended up upgrading the dependency that required the conflicting package. I recommend you start from there. Good luck.

Thank you. It worked for me!

**Comment by kharf** [comment]
Also additionally for others who use go get -u ./...  for updates or renovatebot, which makes use of this command;
go get -u ./... will just hang indefinitely. Even with -x you won't see anything there indicating the issue. Only after I used dependabot, which apparently makes use of single "go gets" per dependency, I saw the error: "can't find reason for requirement on github.com/google/pprof ...."

After updating pprof manually, followed by a "go mod tidy", go get -u ./... works.

This was on go 1.20.

**Comment by thomaspeugeot** [comment]
bug met

```
go version
> go version go1.20.2 darwin/amd64
git clone https://github.com/fullstack-lang/ladybugsim.git
cd ladybugsim
go get -d github.com/fullstack-lang/gongsvg@HEAD github.com/fullstack-lang/gong@HEAD github.com/fullstack-lang/gongdoc@HEAD github.com/fullstack-lang/gongsim@HEAD
panic: internal error: can't find reason for requirement on golang.org/x/image@v0.0.0-20210628002857-a66eb6448b8d

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0xc0006a37b8, 0x12}, {0xc00002ae10, 0x22}})
        /usr/local/go/src/cmd/go/internal/modget/get.go:1760 +0xd4
cmd/go/internal/modget.(*resolver).updateBuildList(0xc0000f8000, {0x1744938, 0xc000026068}, {0x0, 0x0, 0x0})
        /usr/local/go/src/cmd/go/internal/modget/get.go:1765 +0x54c
cmd/go/internal/modget.(*resolver).resolveQueries(0xc0000f8000, {0x1744938, 0xc000026068}, {0xc000074260, 0x4, 0xc000113d70?})
        /usr/local/go/src/cmd/go/internal/modget/get.go:1243 +0x1b1
cmd/go/internal/modget.runGet({0x1744938, 0xc000026068}, 0xc0000285d0?, {0xc0000220a0, 0x4, 0x4})
        /usr/local/go/src/cmd/go/internal/modget/get.go:314 +0x408
main.invoke(0x1a5f260, {0xc000022080, 0x6, 0x6})
        /usr/local/go/src/cmd/go/main.go:225 +0x3d9
main.main()
        /usr/local/go/src/cmd/go/main.go:179 +0x7ce
```

**Comment by anacrolix** [comment]
@kharf I've seen the infinite hang a lot lately, I commented in https://github.com/golang/go/issues/55955#issuecomment-1494181619 thinking it was related to that other issue.

**Comment by bcmills** [comment]
I've mailed [CL 471595](https://go.dev/cl/471595) with a proposed fix. Folks who are affected by this bug: please try out a build with that change (either now or after it is merged) and let me know if any problems remain.

That change may also fix some of the “infinite hang” problems, since it fixes a source of errors in the build list. (`go get` iterates until the build list stabilizes, and I observed some infinite loops in testing caused by pruned dependencies being reintroduced or repeatedly re-eliminated.)

**Comment by francislavoie** [comment]
I don't have any experience with gerrit. Are there instructions somewhere I could follow? Or is there a branch on GitHub I can pull from?

One key point for Caddy in particular is that we can't build from a Go version which reports 1.21 because `quic-go` has custom `crypto/tls` code and only has implementations ready for stable releases. Is there a simple way of making Go report 1.20 for purposes of testing with these changes?

**Comment by bcmills** [comment]
> I don't have any experience with gerrit. Are there instructions somewhere I could follow? Or is there a branch on GitHub I can pull from?

There is a `Download patch` button at the top of the diffs:
![image](https://user-images.githubusercontent.com/5200974/234045513-2520bd8b-ea11-4368-bfd9-cd1f2df94ecc.png)


> Is there a simple way of making Go report 1.20 for purposes of testing with these changes?

Unfortunately no. You could perhaps patch `go-quic` to force it to try building (or add a tag that you could set to force it to try building)?

You could also run `go get` using a `go` built from head, but then do the actual `go build` or `go install` with an older release. (The `go.mod` file produced by the updated `go get` should still be interpreted correctly by previous releases.)

**Comment by mholt** [comment]
Thank you for implementing a hopeful patch/fix! :tada: :blue_heart: We are very grateful for this progress.

> Unfortunately no.

As in, impossible? We're not sure how to test this any other way, since if we mess with the dependency that caused the bug in the first place we can't be sure it's a good test of the patch.

**Comment by bcmills** [comment]
@mholt, you could patch the Go toolchain to pretend that it's 1.20, but in this sort of case — especially since https://github.com/quic-go/quic-go/issues/2727 was filed back in 2020 — it's kind of “you break it, you get to keep all the pieces”. Reaching into release-specific internals of the standard library is definitely not something the Go project supports. 🤷‍♂️

**Comment by mholt** [comment]
Totally understand :100: That's probably acceptable since it's just a local change in a disposable environment for testing a PR. We'll see if we get a chance here soon to try it.

**Comment by anacrolix** [comment]
@bcmills The CL fixes this issue for me too. Specifically in https://github.com/anacrolix/torrent/issues/421#issuecomment-1496917847 I saw
```
~/src/rms-torrent [2]% go1.19 get github.com/anacrolix/torrent@latest
panic: internal error: can't find reason for requirement on github.com/mattn/go-sqlite3@v1.14.16

goroutine 1 [running]:
cmd/go/internal/modget.(*resolver).updateBuildList.func1({{0x14000029640?, 0x140001b37a0?}, {0x140001a2fa8?, 0x14000112b20?}})
	/Users/anacrolix/src/go1.19/src/cmd/go/internal/modget/get.go:1760 +0xf8
cmd/go/internal/modget.(*resolver).updateBuildList(0x14000412000, {0x102b7ae90, 0x140001a2000}, {0x0, 0x0, 0x0})
	/Users/anacrolix/src/go1.19/src/cmd/go/internal/modget/get.go:1765 +0x450
cmd/go/internal/modget.(*resolver).resolveQueries(0x14000412000, {0x102b7ae90, 0x140001a2000}, {0x140001b4048, 0x1, 0x102b601e0?})
	/Users/anacrolix/src/go1.19/src/cmd/go/internal/modget/get.go:1243 +0x180
cmd/go/internal/modget.runGet({0x102b7ae90, 0x140001a2000}, 0x140001be420?, {0x1400019e1a0, 0x1, 0x1})
	/Users/anacrolix/src/go1.19/src/cmd/go/internal/modget/get.go:314 +0x38c
main.invoke(0x102e6e2c0, {0x1400019e190, 0x2, 0x2})
	/Users/anacrolix/src/go1.19/src/cmd/go/main.go:225 +0x2f8
main.main()
	/Users/anacrolix/src/go1.19/src/cmd/go/main.go:179 +0x7ac
```
And now I get 
```
~/src/rms-torrent [2]% go get github.com/anacrolix/torrent@latest
go: upgraded github.com/anacrolix/torrent v1.49.0 => v1.49.1
```

**Comment by anacrolix** [comment]
I think I still get the infinite hang I mentioned at https://github.com/golang/go/issues/56494#issuecomment-1496919121.

**Comment by bcmills** [comment]
@anacrolix, if you're getting an infinite hang try running `go get` with `-v` and/or sending `SIGQUIT`, and please do file a separate issue.

**Comment by bcmills** [comment]
The original steps to reproduce this issue no longer do, but I'm quite confident that it is a symptom of #55955 and will be fixed by https://go.dev/cl/471595.

**Comment by bcmills** [comment]
Duplicate of #55955

**Comment by icio** [comment]
Thank you for all your great work, @bcmills :)
